tests/integration/sandbox.py

```diff
@@ -718,6 +722,188 @@ def restart(self):
+
+    def log_command(self, thread: int, ops: List[ElleOp]):
+        try:
+            self.index[thread] += 1
+        except KeyError:
+            self.index[thread] = 1
+
+        output = self.generate_command_output(thread, ops, "invoke")
+        self.lock.acquire()
+        self.logfile.write(output)
+        self.logfile.write("\n")
+        self.lock.release()
+
+    @staticmethod
+    def generate_results_value(results: typing.List, ops: typing.List[ElleOp]) -> str:
+        val = "["
+        for i in range(len(ops)):
+            if i != 0:
+                val += " "
+            if ops[i].op_type == "append":
+                val += str(ops[i])
+            elif ops[i].op_type == "read":
+                tmp = " ".join([x.decode('utf-8') for x in results[i]])
+                val += f"[:r {ops[i].key} [{tmp}]]"
+            else:
+                raise Exception(f"unknown op type {ops[i].op_type}")
+        val += "]"
+        return val
+
+    def generate_result_output(self, thread: int, result: str, results: typing.Optional[List], ops: List[ElleOp]):
+        output: str
+        if result == "ok":
+            output = self.generate_results_value(results, ops)
+            return f"{{:index {self.index[thread]} :process {thread} :type :{result}, :value {output}}}"
+        elif result == "fail":
+            output = self.generate_ops_value(ops)
+            return f"{{:index {self.index[thread]} :process {thread} :type :{result}, :value {output}}}"
+        else:
+            raise Exception(f"unknown result value {result}")
+
+    def log_result(self, thread, result: str, results: typing.Optional[List], ops: List[ElleOp]):
+        # if this throws an exception, we have a failure elsewhere
+        self.index[thread] += 1
+
+        output = self.generate_result_output(thread, result, results, ops)
+        self.lock.acquire()
+        self.logfile.write(output)
+        self.logfile.write("\n")
+        self.lock.release()
+
+    @staticmethod
+    def map_addresses_to_clients(clusters: List[Cluster]) -> typing.Mapping[str, redis.Redis]:
+        mapped_clients: typing.Dict[str, redis.Redis] = {}
+        default_set = False
+        for c in clusters:
+            for n in c.nodes.values():
+                mapped_clients[n.address] = n.client
+                if not default_set:
+                    mapped_clients["default"] = n.client
+                    default_set = True
+        return mapped_clients
+
+    def get_next_value(self) -> int:
+        with self.lock:
+            ret = self.value
+            self.value += 1
+            return ret
+
+
+class ElleWorker(object):
+    def __init__(self, elle: Elle, clients: typing.Dict[str, redis.Redis]):
+        self.clients = clients
+        self.elle = elle
+        self.id = threading.get_ident()
+
+    # might not want generate_ops, perhaps each multi should just be a single append/read of same key
```

Comment:
I don't see any difference between single append vs single append + read. Read version is probably better, gives more opportunities to catch a bug. So, I guess we should make it append + read.

Reply:
ok, I think thats what it is now, right?

Reply:
I was just replying to your comment there :) Is that comment invalid now? Should we delete it? orshould we simplify this function?

Reply:
I agree that append/read is what we want.